Core/src/Surfaces/BoundaryTolerance.cpp

```diff
@@ -0,0 +1,198 @@
+    : m_variant{std::move(variant)} {}
+
+bool BoundaryTolerance::isInfinite() const {
+  return holdsVariant<Infinite>();
+}
+
+bool BoundaryTolerance::isNone() const {
+  return holdsVariant<None>();
+}
+
+bool BoundaryTolerance::hasAbsoluteBound(bool isCartesian) const {
+  return holdsVariant<None>() || holdsVariant<AbsoluteBound>() ||
+         (isCartesian && holdsVariant<AbsoluteCartesian>());
+}
+
+bool BoundaryTolerance::hasAbsoluteCartesian() const {
+  return holdsVariant<AbsoluteCartesian>();
+}
+
+bool BoundaryTolerance::hasAbsoluteEuclidean() const {
+  return holdsVariant<AbsoluteEuclidean>();
+}
+
+bool BoundaryTolerance::hasChi2Bound() const {
+  return holdsVariant<Chi2Bound>();
+}
+
+bool BoundaryTolerance::hasTolerance() const {
+  if (isInfinite()) {
+    return true;
+  }
+
+  if (isNone()) {
+    return false;
+  }
+
+  if (auto absoluteBound = getVariantPtr<AbsoluteBound>();
+      absoluteBound != nullptr) {
+    return absoluteBound->tolerance0 != 0. || absoluteBound->tolerance1 != 0.;
+  }
+
+  if (auto absoluteCartesian = getVariantPtr<AbsoluteCartesian>();
+      absoluteCartesian != nullptr) {
+    return absoluteCartesian->tolerance0 != 0. ||
+           absoluteCartesian->tolerance1 != 0.;
+  }
+
+  if (auto absoluteEuclidean = getVariantPtr<AbsoluteEuclidean>();
+      absoluteEuclidean != nullptr) {
+    return absoluteEuclidean->tolerance != 0.;
+  }
+
+  if (auto chi2Bound = getVariantPtr<Chi2Bound>(); chi2Bound != nullptr) {
+    return chi2Bound->maxChi2 != 0.;
+  }
+
+  assert(false && "Unsupported tolerance type");
+  return false;
+}
+
+BoundaryTolerance::AbsoluteBound BoundaryTolerance::asAbsoluteBound(
+    bool isCartesian) const {
+  if (isNone()) {
+    return AbsoluteBound{0., 0.};
+  }
+
+  if (isCartesian && hasAbsoluteCartesian()) {
+    const auto& cartesian = getVariant<AbsoluteCartesian>();
+    return AbsoluteBound{cartesian.tolerance0, cartesian.tolerance1};
+  }
+
+  return getVariant<AbsoluteBound>();
+}
+
+const BoundaryTolerance::AbsoluteCartesian&
+BoundaryTolerance::asAbsoluteCartesian() const {
+  return getVariant<AbsoluteCartesian>();
+}
+
+const BoundaryTolerance::AbsoluteEuclidean&
+BoundaryTolerance::asAbsoluteEuclidean() const {
+  return getVariant<AbsoluteEuclidean>();
+}
+
+const BoundaryTolerance::Chi2Bound& BoundaryTolerance::asChi2Bound() const {
+  return getVariant<Chi2Bound>();
+}
+
+std::optional<BoundaryTolerance::AbsoluteBound>
+BoundaryTolerance::asAbsoluteBoundOpt(bool isCartesian) const {
+  return hasAbsoluteBound(isCartesian)
+             ? std::optional(asAbsoluteBound(isCartesian))
+             : std::nullopt;
+}
+
+bool BoundaryTolerance::isTolerated(
+    const Vector2& distance,
+    const std::optional<SquareMatrix2>& jacobianOpt) const {
+  if (isInfinite()) {
+    return true;
+  }
+
+  if (isNone()) {
+    return distance == Vector2::Zero();
+  }
+
+  if (auto absoluteBound = getVariantPtr<AbsoluteBound>();
```

Comment:
Nitpick:

```suggestion
  if (const auto* absoluteBound = getVariantPtr<AbsoluteBound>();
```

also applies below.